File: conway.py
#!/usr/bin/env python

import unittest


class Grid(object):

    def __init__(self, cell_list):
        self.cells = cell_list


    def is_alive(self, x, y):
        d = {True:1, False:0}
        return d[(x,y) in self.cells]


    def num_live_neighbors(self, x, y):
        return sum(self.is_alive(i,j) 
            for i in range(x-1, x+2)
            for j in range(y-1, y+2)) - self.is_alive(x,y)
        

    def lives(self, x, y):
        live_now = dict((key, False) for key in range(9))
        dead_now = dict((key, False) for key in range(9))
        live_now[2] = True
        live_now[3] = True
        dead_now[3] = True
        survives = {0:dead_now, 1:live_now}
        return survives[self.is_alive(x,y)][self.num_live_neighbors(x,y)]


    def tick(self):
        new_cells = []
        x_list = [x for (x,y) in self.cells]
        y_list = [y for (x,y) in self.cells]
        for i in range(min(x_list)-1, max(x_list)+2):
            for j in range(min(y_list)-1, max(y_list)+2):
                if self.lives(i,j):
                    new_cells.append((i,j))
        self.cells = new_cells


    def __repr__(self):
        n=4
        rep = [ ['.']*n ] * n
        rep = [ ['.']*n for i in range(n) ]
        for (x,y) in self.cells: rep[x][y] = 'X'
        return ''.join(''.join(cell for cell in row)+'\n' for row in rep)



class TestGrid(unittest.TestCase):
    cell_list = [(2,0), (2,1), (2,2)]
    '''
     t0     t1    t2
    ..x..  ..... ..x..
    ..x..  .xxx. ..x..
    ..x..  ..... ..x..
    .....  ..... .....
    '''

    def setUp(self):
        self.grid = Grid(self.cell_list)

    def test_is_alive(self):
        self.assertEqual(self.grid.is_alive(2,0), 1)
        self.assertEqual(self.grid.is_alive(3,2), 0)

    def test_num_live_neighbors(self):
        self.assertEqual(self.grid.num_live_neighbors(0,0), 0)
        self.assertEqual(self.grid.num_live_neighbors(1,1), 3)
        self.assertEqual(self.grid.num_live_neighbors(5,5), 0)
        self.assertEqual(self.grid.num_live_neighbors(2,2), 1)


    def test_lives(self):
        self.assertFalse(self.grid.lives(0,0))
        self.assertTrue(self.grid.lives(1,1))
        self.assertFalse(self.grid.lives(2,2))
        self.assertTrue(self.grid.lives(2,1))


    def test_tick(self):
        tick_1 = [(1,1), (2,1), (3,1)]
        self.grid.tick()
        self.assertEqual(sorted(self.grid.cells), sorted(tick_1))
        tick_2 = [(2,0), (2,1), (2,2)]
        self.grid.tick()
        self.assertEqual(sorted(self.grid.cells), sorted(tick_2))


def visual_test(cell_list):
    grid = Grid(cell_list)
    print grid
    for i in range(5):
        grid.tick() 
        print grid



if False and __name__ == '__main__':
    visual_test([(0,1), (1,1), (2,1)]) # blinker
    visual_test([(0,0), (2,0), (2,1), (1,1), (1,2)]) # glider
    unittest.main()

File: conway.js
// Conway's Game of Life in javascript with graphics by d3js //
//
// triv


grid = {}
grid.__init__ = function(cell_list){ 
    grid.cells = cell_list }


grid.is_alive = function(x, y){
    var d = {true:1, false:0}
    return d[in_(grid.cells, [x,y])] }


grid.num_live_neighbors = function(x, y){
    res = 0
    for (var i=n+x-1; i<n+x+2; i++)
        for (var j=n+y-1; j<n+y+2; j++)
            res += grid.is_alive(i%n,j%n)
    return res - grid.is_alive(x%n,y%n)
}


grid.lives = function(x, y){
    var live_now = {
      0:false, 1:false, 2:true, 3:true, 4:false, 5:false, 6:false, 7:false, 8:false}
    var dead_now = {
      0:false, 1:false, 2:false, 3:true, 4:false, 5:false, 6:false, 7:false, 8:false}
    var survives = {0:dead_now, 1:live_now}
    return survives[grid.is_alive(x%n,y%n)][grid.num_live_neighbors(x,y)]
}


grid.tick = function(){
    var new_cells = []
    var x_list = []
    var y_list = []
//    console.log('new_cells: '+new_cells)
    for (var i in grid.cells){
        x_list.push(grid.cells[i][0])
        y_list.push(grid.cells[i][1])
    }
//    console.log('x_list: '+x_list)
    x_min = Math.min.apply(Math, x_list)-1
    x_max = Math.max.apply(Math, x_list)+2
    y_min = Math.min.apply(Math, y_list)-1
    y_max = Math.max.apply(Math, y_list)+2
//    console.log('y_max: '+y_max)

    for (var i=x_min; i<x_max; i++)
        for (var j=y_min; j<y_max; j++){
            if (grid.lives(i,j))
                new_cells.push([i%n,j%n])
            }
    grid.cells = dedupe(new_cells.map(normalize))
//    console.log('new_cells: '+new_cells)
    set_active_cells(grid.cells)
}

// That is all the Conway calculations.  Everything else is display.


function normalize(pair){
    // add the modulus until both x,y are positive.
//    return pair.map(positive_mod)  NO
    x=pair[0]
    y=pair[1]
    while (x<0) x+=n
    while (y<0) y+=n
    return [x,y]
}

function positive_mod(x, mod){
    // add the modulus until x is positive.
    while (x<0) x+=mod
    return x
}
// TODO
// positive_mod is now a general function for util and can get the effect
// on the pair by saying pair.map(positive_mod)
// NO. find out why.



function draw_grid(mat) {
    table = d3.select("svg")
    table.selectAll("rect").remove()

    table
    .selectAll("rect")
    .data(mat)
    .enter().append("rect")

    .on("click",  toggle_class) 
    .text(function(d){return '['+d+']'}) // save coords as text.
    .attr("x", function(d){return x(d[0])})
    .attr("y", function(d){return y(d[1])})
    .attr("height", 10)
    .attr("width", 10)
}

n = 10
n = 9
n = 8
n = 20
n = 30

// d3 range funcs.
var x = d3.scale.linear()
    .domain([0,n-1])
    .range([0,n*10])

var y = d3.scale.linear()
    .domain([0,n-1])
    .range([0,n*10])


function get_active_cells(){
    var coord_list = []

    // Extract coords as text.
    function ea_func(idx, elt){ 
        coord_list.push( eval(elt.childNodes[0].data ))
    }
    $('rect.active').each(ea_func)
    grid.__init__(coord_list)
    return coord_list
}


function set_active_cells(cells){
    // Updating d3 display.
    // Update data can be any length but it will be applied to the first
    // update.length items.
    // So in general we want the update data list to be same length as
    // original list.
    //
    var d = {true:'active', false:'inactive'}
    var new_data = mat.map(function(pair){ return d[in_(cells, pair)] })
    // new_data is a list of strings 'active'/'inactive'

    d3.select("svg")
    .selectAll("rect")
    .data(new_data)
    .attr("class", function(d){ return d })

}


go_func = function (){ go_stop = setInterval(grid.tick, 500) }
stop_func = function (){ clearInterval(go_stop)}

function buttons(){
    // Grid data is a list of coordinates.
   mat = []
    for (var i=0; i<n; i++)
        for (var j=0; j<n; j++)
            mat.push([i,j])
     

    draw_grid(mat)
    $('#go_button').click(function (){ get_active_cells(); go_func() })
    $('#stop_button').click( stop_func )
    $('#step_button').click(function (){ get_active_cells(); grid.tick() })
    $('#reset_button').click(function (){ stop_func(); draw_grid(mat) })
}



function other_test(){
    cell_list = [[0,1], [1,1], [2,1]]
    cell_list = [[0,0], [1,1], [2,2], [3,3], [2,3]]
    cell_list = [[4,5], [4,6], [4,7]]
    grid.__init__(cell_list)
}

